### PR TITLE
send transactions to peers concurrently & preemptively

### PR DESCRIPTION
**Pull request by levikobi**
First I want to say that I really enjoy your courses, and learned a lot from them.

While watching the blockchain class, I noticed that the worker responsible for sending transactions to the peers is doing it in a synchronous way, which isn’t going to scale well for a large amount of peers & transactions. So, I figured we can at least update the peers simultaneously.
In addition, shareTxOperations() might block forever if a node stalls the response, without a way for the current node to continue, or shut down.

I also added a context to the “send” function under worker/network to be able to cancel network calls. To my understanding, if a bad node decides to stall the private api sync requests it could possibly halt the entire blockchain in the current implementation. Adding the context is not a complete solution, which requires a punishment system for bad nodes, but it could be a start.

A context.Todo() was used in the other NetRequest functions, since I didn’t know whether you will go ahead with this change or not. If you do, I can update the rest.

### CODE REVIEW

**Comment by ardan-bkennedy** [comment]
Thanks for this. I review it but I'm not sure the extra concurrency is necessary here. This code is a reference implementation and the added complexity isn't really necessary. I'll review this when I can

File: foundation/blockchain/worker/sharetx.go [hunk 1]
```diff
@@ -20,15 +25,30 @@ func (w *Worker) shareTxOperations() {
 	w.evHandler("worker: shareTxOperations: G started")
 	defer w.evHandler("worker: shareTxOperations: G completed")
 
-	for {
-		select {
-		case tx := <-w.txSharing:
-			if !w.isShutdown() {
-				w.state.NetSendTxToPeers(tx)
+	var wg sync.WaitGroup
```
Comment (ardan-bkennedy) [inline foundation/blockchain/worker/sharetx.go:28]: All of this is already running as a Goroutine. The Concurrency code is wrapped around this call. So no need for the extra concurrency here.

File: foundation/blockchain/state/network.go [hunk 2]
```diff
@@ -36,26 +39,39 @@ func (s *State) NetSendBlockToPeers(block database.Block) error {
 }
 
 // NetSendTxToPeers shares a new block transaction with the known peers.
-func (s *State) NetSendTxToPeers(tx database.BlockTx) {
+func (s *State) NetSendTxToPeers(ctx context.Context, tx database.BlockTx) {
```
Comment (ardan-bkennedy) [inline foundation/blockchain/state/network.go:42]: I hear what you are saying, but this is really not necessary. In a production version you would want to send this to your local group of peers quickly, that is true. Adding the concurrency here only makes the code harder to read and comprehend for new people. This is running in a goroutine, just these calls are serialized.

Comment (ardan-bkennedy) [reply foundation/blockchain/state/network.go:42]: Also, look at how I applied concurrency. These functions perform the work they need to and the concurrency is coded outside these functions. Look at worker.go.  You really want to separate these things. For the POW code, that required extra concurrency, but notice again the concurrency code does it's thing and the functions that code calls don't know anything about how concurrency.

Comment (levikobi) [reply foundation/blockchain/state/network.go:42]: I think I understand what you're saying.
runPowOperation() & runPoaOperation() are responsible for concurrency as opposed to powOperations() & poaOperations(), as they are listening to the cancelMining channel.

The difference, in my opinion, is that NetSendTxToPeers() does not listen to any "cancelMining" equivalent. It is a blocking call, that may block forever as things are right now, and you won't be able to cancel it (hence the new goroutine - to be able to cancel NetSendTxToPeers).

But anyway I completely understand your point as this adding extra complexity, which can be avoided in a reference implementation.

Thanks for the review!

**Pull request closed without merging**